src/transaction/utils.ts

```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
   };
 }
 
+/**
+ * Autofills fields in a transaction.
+ *
+ * @param {Transaction} tx A transaction to autofill fields.
+ * @param {Client} client A client.
+ * @returns {Transaction} An autofilled transaction.
+ */
+async function autofillTransaction(
+  tx: Transaction,
+  client: Client
+): Promise<Transaction> {
+  validateAccountAddress(tx, 'Account', 'SourceTag');
+  if (tx['Destination'] != null) {
+    validateAccountAddress(tx, 'Destination', 'DestinationTag');
+  }
+
+  // DepositPreauth:
+  convertToClassicAddress(tx, 'Authorize');
+  convertToClassicAddress(tx, 'Unauthorize');
+  // EscrowCancel, EscrowFinish:
+  convertToClassicAddress(tx, 'Owner');
+  // SetRegularKey:
+  convertToClassicAddress(tx, 'RegularKey');
+
+  setTransactionFlagsToNumber(tx);
+
+  if (tx.Sequence == null) {
+    const sequence = await getNextValidSequenceNumber(tx.Account, client);
+    tx.Sequence = sequence;
+  }
+
+  if (tx.Fee == null) {
+    tx.Fee = await calculateFeePerTransactionType(tx, client);
+  }
+
+  if (tx.LastLedgerSequence == null) {
+    const ledgerSequence = await getLatestValidatedLedgerSequence(client);
+    tx.LastLedgerSequence = ledgerSequence + LEDGER_OFFSET;
+  }
+
+  return tx;
+}
+
+function validateAccountAddress(tx: Transaction, accountField: string, tagField: string): void {
+  // if X-address is given, convert it to classic address
+  const { classicAccount, tag } = getClassicAccountAndTag(tx[accountField]);
+  tx[accountField] = classicAccount;
+
+  if (tag !== null) {
+    if (tx[tagField] && tx[tagField] !== tag) {
+      throw new ValidationError(
+        `The ${tagField}, if present, must match the tag of the ${accountField} X-address`
+      );
+    }
+    tx[tagField] = tag;
+  }
+}
+
+function convertToClassicAddress(tx: Transaction, fieldName: string): void {
+  const account = tx[fieldName];
+  if (typeof account === "string") {
+    const { classicAccount } = getClassicAccountAndTag(account);
+    tx[fieldName] = classicAccount;
+  }
+}
+
+async function getNextValidSequenceNumber(
+  account: string,
+  client: Client
+): Promise<number> {
+  const request: AccountInfoRequest = {
+    command: 'account_info',
+    account,
+  };
+  const data = await client.request(request);
+  return data.result.account_data.Sequence;
+}
+
+async function calculateFeePerTransactionType(
+  transaction: Transaction,
+  client?: Client
+): Promise<string> {
+  let netFee;
+  if (client == null) {
+    netFee = 10; // 10 drops
+  } else {
+    netFee = await client.getFee(); // Usually 0.00001 XRP (10 drops)
+  }
+
+  let baseFee = netFee;
+
+  // EscrowFinish Transaction with Fulfillment
+  if (transaction.TransactionType === 'EscrowFinish') {
+    if (transaction.Fulfillment != null) {
+      const fulfillmentBytesSize = Math.ceil(
+        transaction.Fulfillment.length / 2
+      );
+      // 10 drops × (33 + (Fulfillment size in bytes / 16))
+      baseFee = Math.ceil(netFee * (33 + fulfillmentBytesSize / 16));
+    }
```

Comment:
this conditional nesting is unnecessary since there is nothing that happens if the outer condition is true but the inner condition is false. as a result you should unnest this. here's one suggestion

```typescript
const isFulfillment = transaction.TransactionType === 'EscrowFinish' && transaction.Fulfillment != null
if (isFullfillment) {
  const fulfillmentBytesSize = Math.ceil(.....)
  ....
}
```

Reply:
Yup you're right, this totally went over my head copying our Python implementation. 😄